### PR TITLE
examples/window: use softbuffer

### PR DESCRIPTION
**Pull request by dhardy**
See also #2350, https://github.com/john01dav/softbuffer/pull/20.

- [ ] Tested on all platforms changed
- [ ] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [ ] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [ ] Created or updated an example program if it would help users understand this functionality
- [ ] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

### CODE REVIEW

File: examples/window.rs [hunk 1]
```diff
@@ -17,17 +18,29 @@ fn main() {
...
 
         match event {
+            Event::RedrawRequested(window_id) if window_id == graphics_context.window().id() => {
+                let (width, height) = {
+                    let size = graphics_context.window().inner_size();
+                    (size.width, size.height)
+                };
+                let colour = 0xffee36;
+                let mut buffer = Vec::new();
+                buffer.resize(width as usize * height as usize, colour);
```
Comment (filnet) [inline examples/window.rs:35]: should use `Vec::with_capacity()` to create a properly sized buffer upfront.

Comment (kchibisov) [reply examples/window.rs:35]: If we're talking about optimizing. Doing `vec![color; (width * height) as usize]` is the way to go.

Comment (dhardy) [reply examples/window.rs:35]: @filnet it only allocates once so no difference. @kchibisov is of course neater.

Comment (filnet) [reply examples/window.rs:35]: > @filnet it only allocates once so no difference.

I learned something today.

**Comment by maroider** [comment]
The X11-only CI appears to be failing due to a somehow missing `waland-client` system library. Not sure what's up with that.

**Comment by dhardy** [comment]
`softbuffer` has this dependency:

> wayland-client = {version = "0.29", features = ["use_system_lib"], default_features = false}

I tried adding `libwayland-client0` as a dependency... though since `winit` itself already uses `wayland-client` with this feature I'm doubtful.

**Comment by dhardy** [comment]
No surprise, that didn't help:

> libwayland-client0 is already the newest version (1.18.0-1).

Strangely, its the `x11` build that failed.

Running `cargo test --no-run --verbose --target x86_64-unknown-linux-gnu --no-default-features --features x11` (what the command reduces to) on my machine (running X11) works.

**Comment by maroider** [comment]
> Strangely, its the x11 build that failed.

Indeed. It's quite strange. Could it be that there's some side-effect `softbuffer` inadvertently relies on when we enable our wayland deps?

**Comment by i509VCB** [comment]
Personally I'd be a little skeptical of using softbuffer as an example as it has some bugs. And there is also the maintenance issue as several issues and pull requests have been opened in the last 2 months without any responses.

I do recall we had a `winit-blit` project, but that never really took off.

**Comment by madsmtm** [comment]
We've forked [`swbuf`](https://github.com/rust-windowing/swbuf) under @rust-windowing, so we'll probably want to use that instead - I'll close this for now, esp. also because `swbuf` doesn't support all of `winit`'s platforms yet (Android and iOS missing), but free to reopen if you update it to use that

**Comment by dhardy** [comment]
Good to hear that there's a maintained fork. If you'd rather see `swbuf` support more platforms first then I guess there's no point updating this PR for now.

**Comment by MarijnS95** [comment]
Randomly found this (branch deletion poked the PR), iOS/MacOS are now supported and Android will be soon: https://github.com/rust-windowing/softbuffer/pull/288

Seems like `winit` now already uses `softbuffer` in some examples, but disabled it on these platforms for "some reason" :/

**Pull request closed without merging**